build-logic/jvm/src/main/kotlin/gradlebuild.unittest-and-compile.gradle.kts

```
@@ -127,23 +138,67 @@ fun enforceCompatibility(gradlebuildJava: UnitTestAndCompileExtension) {
+    val versionProviders = listOf(
+        versionProvider(usedForStartup, 6),
+        versionProvider(usedInWrapper, 6), // TODO: Should be 8
+        versionProvider(usedInWorkers, 8),
+        versionProvider(usedInClient, 8),
+        versionProvider(usedInDaemon, 8),
+    )
+
+    val iterator = versionProviders.iterator()
+    var state = iterator.next()
+    while (iterator.hasNext()) {
+        state = state.zip(iterator.next()) { left , right ->
+            if (left.isPresent) {
+                if (right.isPresent) {
+                    Optional.of(minOf(left.get(), right.get()))
+                } else {
+                    left
+                }
+            } else {
+                right
+            }
+        }
+    }
+
+    return state.map { it.orElseThrow {
+        GradleException("No target JVM version configured. Specify a target platform on ${UnitTestAndCompileExtension::class.java.simpleName}")
+    }}
+}
+
+fun versionProvider(booleanProperty: Provider<Boolean>, version: Int): Provider<Optional<Int>> =
+    booleanProperty.map {
+        if (it) {
+            Optional.of(version)
+        } else {
+            Optional.empty()
+        }
+    }
+
 fun enforceJavaCompatibility(targetVersion: Provider<Int>, useRelease: Provider<Boolean>) {
     tasks.withType<JavaCompile>().configureEach {
         // Set the release flag is requested.
         // Otherwise, we set the source and target compatibility in the afterEvaluate below.
-        options.release = useRelease.flatMap { doUseRelease -> targetVersion.filter { doUseRelease } }
+        options.release = useRelease.zip(targetVersion) { doUseRelease, target -> if (doUseRelease) { target } else { null } }
 
         // If we are targeting Java < 8, we need to use a different compiler,
         // since compilers will only cross-compile down to a certain version.
-        javaCompiler = javaToolchains.compilerFor {
-            languageVersion = targetVersion.flatMap {
-                if (it >= 8) {
-                    // The toolchain on the project is able to target JDK >= 8
-                    java.toolchain.languageVersion
-                } else {
-                    // To compile Java 6 and 7 sources, we need an older compiler
-                    // We choose 11 since it supports both of these versions.
-                    provider { JavaLanguageVersion.of(11) }
+        javaCompiler = targetVersion.flatMap { version ->
+            if (version >= 8) {
+                // The build jvm (Java 17) is able to target JDK >= 8
+                javaToolchains.compilerFor(java.toolchain)
+            } else {
+                // To compile Java 6 and 7 sources, we need an older compiler.
+                // We choose 11 since it supports both of these versions.
+                javaToolchains.compilerFor {
+                    languageVersion = JavaLanguageVersion.of(11)
                 }
             }
```

💅 I would move compiler lookups out of the `flatMap` closure to avoid capturing extra stuff there, but this is a very minor thing that shouldn't cause any problems, at least with the current implementation of CC or task dependency discovery.

---

I suggest extracting the compiler providers into variables:
```suggestion
        val defaultCompiler = javaToolchains.compilerFor(java.toolchain)
        // If we are targeting Java < 8, we need to use a different compiler,
        // since compilers will only cross-compile down to a certain version.
        val legacyCompiler = javaToolchains.compilerFor {
            // We choose 11 since it supports both 6 and 8.
            languageVersion = JavaLanguageVersion.of(11)
        }
        javaCompiler = targetVersion.flatMap { version ->
            if (version >= 8) {
                // The build jvm (Java 17) is able to target JDK >= 8
                defaultCompiler
            } else {
                // To compile Java 6 and 7 sources, we need an older compiler.
                legacyCompiler
            }
        }
```

---

I see, makes sense. Will update to this